[GEOS-10658] - Added OpenLayers example using vectortiles in WMS for vector tiles tutorial (#6199)
geoserver/geoserver: doc/en/user/source/extensions/vectortiles/tutorial.rst

GeoServer can also produce vector tiles in three formats: GeoJSON, TopoJSON, and MapBox Vector (MVT). These are also supported by OpenLayers and other clients.

> .. warning:: When using vector tiles, be sure to use an up-to-date client. Older clients do not support all vector tiles capabilites and may result in rendering errors. We recommend using the latest version of OpenLayers (Currently v5.3.0).

Reviewer: Why this change?

Author: I have to be fair I am not too sure what happened there. I fixed it anyway

> .. warning:: When using vector tiles, be sure to use an up-to-date client. Older clients do not support all vector tiles capabilites and may result in rendering errors. We recommend using the latest version of OpenLayers (Currently v5.3.0).

Author: Replied above to this

     - This is a human readable JSON format.  Although many geo-spatial applications support GeoJSON datasets, few Vector Tile applications support tiles in this format.  Supported by Open Layers 3.
   * - `TopoJSON <https://github.com/mbostock/topojson/wiki>`_
     - ``application/json;type=topojson``

>      - This is a very complex, but somewhat human readable JSON format that is good for polygon coverages.  It is not a widely supported and very few Vector Tile applications support it.  Suported by Open Layers 3.

Reviewer: Why this change?

Author: I have to be fair I am not too sure what happened there. I fixed it anyway

Reviewer: It still appears to bad in the review version.  Can you check you have everything committed?

Our layer is now ready to be served.

Create OpenLayers application - TMS Vector Tiles
------------------------------------------------

> #. Create a ``www/tms-vectortiles`` directory inside your GeoServer Data Directory.

Reviewer: Why this change?

Author: @bradh, this was to differentiate this example from the one I included. In case users want to follow both and have them in separated folders `tms-vectortiles` `wms-vectortiles`.

Our layer is now ready to be served.

Create OpenLayers application - TMS Vector Tiles
------------------------------------------------

#. Create a ``www/tms-vectortiles`` directory inside your GeoServer Data Directory.

> #. Download the `latest version of OpenLayers <http://openlayers.org/download/>. Download the v<ol-version>-package.zip file`_.

Reviewer 2: The docs build fails with:

```
07:36:54,213 [INFO]      [exec] /home/runner/work/geoserver/geoserver/doc/en/user/source/extensions/vectortiles/tutorial.rst:88: ERROR: Unknown target name: "latest version of openlayers <http://openlayers.org/download/>. download the v<ol-version>-package.zip file"."
```

The link seems to be missing a small bit at  the end, hopefully this is going to fix it.

```suggestion
#. Download the `latest version of OpenLayers <http://openlayers.org/download/>`_. Download the v<ol-version>-package.zip file`_.
```

Author: Thanks

These tiles are being rendered by the OpenLayers client.

Create OpenLayers application - WMS Vector Tiles
------------------------------------------------

> .. note::   Vector tiles requested with WMS allows to retrieve non-cached vector tiles (server side), setting the ``tiled=false`` parameter on the ``getMap`` request. This setting could be particularly useful when serving fast changing source data that should constantly be kept up-to-date for display.

Reviewer: ```suggestion
.. note::   Vector tiles requested with WMS allows retrieving non-cached vector tiles (server side) by setting the ``tiled=false`` parameter on the ``getMap`` request. This setting could be particularly useful when serving fast changing source data that should constantly be kept up-to-date for display.
```

> .. note::   Vector tiles requested with WMS allows to retrieve non-cached vector tiles (server side), setting the ``tiled=false`` parameter on the ``getMap`` request. This setting could be particularly useful when serving fast changing source data that should constantly be kept up-to-date for display.

Reviewer 2: It should really be qualified by the size of the vector tile generated. One needs to pay attention to generate small ones (few attributes, few features) in order for them to be usable.

Also editing on vector tiles is problematic, geometry locations are quantized on a grid, you are never editing the original position, but a "close" one, how close depends on your current zoom level, just mind it will never be exactly the original one. Also, if dealing with lines/polygons crossing the tile borders, they are also clipped, reconstructing them on the client side from all the tiles containing them (good luck finding when to stop, think the coastline of Asia for example) is possible but causes further accuracy degradation.

> .. note::   Vector tiles requested with WMS allows to retrieve non-cached vector tiles (server side), setting the ``tiled=false`` parameter on the ``getMap`` request. This setting could be particularly useful when serving fast changing source data that should constantly be kept up-to-date for display.

Author: Included the qualification statement.

#. Create a ``www/wms-vectortiles`` directory inside your GeoServer Data Directory.

> #. Download the `latest version of OpenLayers <http://openlayers.org/download/>`_.

Reviewer 2: I see two downloads, site and package... I guess you mean the "package" one?
I see "ol.css" in the root of the zip, "ol.js" in the "dist" zip, don't see a ``ol-debug.js`` though.
This part of the instructions should contain a bit more detail.

Author: @aaime, yes, I copied and pasted it from the previous part. Both should be rewritten since the downloadable `.zip` of OpenLayers has changed structure

Reviewer 2: Then please do so :-D

Author: Done, corrected with the right paths to the files

> #. Extract the following files to from the downloaded archive to the directory created in step 1:

Reviewer: ```suggestion
#. Extract the following files from the downloaded archive to the directory created in step 1:
```

   * ``ol.js``
   * ``ol-debug.js``
   * ``ol.css``

#. In a text editor, create a new file with the following content:

   .. code-block:: html
      :emphasize-lines: 5
      
      <!doctype html>
      <html>
      <head>
        <title>Vector tiles</title>
        <script src="ol.js"></script>
        <link rel="stylesheet" href="ol.css">
        <style>
          html, body {
            font-family: sans-serif;
            width: 100%;
          }
          .map {
            height: 500px;
            width: 100%;
          }
        </style>
      </head>

>         <body>

Reviewer: Looks like the formatting is off. Possibly reformat it before incorporating it?

Author: yes it is off

          <h3>Mapbox Protobuf - vector tiles WMS</h3>
          <div class="refresh-container">
          <button id="refresh-button" type="button" onclick="updateFunc();">Refresh/reload cache</button>
          </div>
          <div id="map" class="map"></div>
          <script>
          

>             var layerParams = {'LAYERS': 'opengeo:countries', 'TILED': false, 'format': 'application/vnd.mapbox-vector-tile'};

Reviewer 2: Something odd here, the generate URL ends up having "format" twice, once uppercase, once lowercase, e.g:

```
http://localhost:8080/geoserver/wms?SERVICE=WMS&VERSION=1.3.0&REQUEST=GetMap&FORMAT=image%2Fpng&TRANSPARENT=true&LAYERS=importer%3Ane_10m_roads&TILED=false&format=application%2Fvnd.mapbox-vector-tile&WIDTH=256&HEIGHT=256&CRS=EPSG%3A3857&STYLES=&BBOX=10018754.17139462%2C0%2C20037508.34278924%2C10018754.171394622
``

Author: @aaime yes 'FORMAT' had to be specified in upper case for the property to be overridden
